neo4j/internal/bolt/message_queue.go

```diff
@@ -0,0 +1,206 @@
+package bolt
+
+import (
+	"container/list"
+	"context"
+	"errors"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/db"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/log"
+	"net"
+)
+
+type messageQueue struct {
+	in               incoming
+	out              outgoing
+	callbacks        list.List // List[responseHandler]
+	targetConnection net.Conn
+	protocolVersion  boltVersion
+	err              error
+
+	onNextMessage    func()
+	onNextMessageErr func(error)
+}
+
+func newMessageQueue(
+	target net.Conn,
+	in incoming, out outgoing,
+	onNext func(),
+	onNextErr func(error)) messageQueue {
+
+	return messageQueue{
+		in:               in,
+		out:              out,
+		callbacks:        list.List{},
+		targetConnection: target,
+		onNextMessage:    onNext,
+		onNextMessageErr: onNextErr,
+	}
+}
+
+func (q *messageQueue) appendHello(protocolVersion boltVersion, hello, token map[string]any,
+	helloHandler responseHandler, logonHandler responseHandler) {
+	q.protocolVersion = protocolVersion
+	q.out.appendHello(hello)
+	q.enqueueCallback(helloHandler)
+	if q.protocolVersion.greaterThanOrEqual(5, 1) {
+		q.out.appendLogon(token)
+		q.enqueueCallback(logonHandler)
+	}
```

Comment:
I think it would make for a cleaner separation of concerns if the queue was more (ideally fully) protocol version agnostic.

In this particular case there could be an `appendHello` method on the queue and `bolt.Connect` is responsible for deciding whether logon needs to be enqueued or not.

Reply:
I agree and had the same thought yesterday evening!
Keeping protocol version-based logic here would make things too complicated (esp. for bolt4 and ROUTE)